### PR TITLE
TT-613 limit number of records which pump pulls from redis list GetAndDeleteSet

### PR DESCRIPTION
**Pull request by asoorm**
Pending full QA

Further addresses: https://github.com/TykTechnologies/tyk-pump/issues/185

Traditionally, we pull the entire analytics list from Redis and delete the set. This is problematic for several reasons:

* If pump fails, **all** analytics are lost after TTL on analytics key expires.
* If multiple pumps are awaiting analytics, some pumps will get no records - so load is unevenly distributed
* If data-sink is slow, then analytics data elements build up in Redis. The pump will then pull ALL elements from the Redis List, and then try to ship all analytics - causing further pressure on the data-sink. Imagine sending 500k analytics records at 1kb each to ElasticSearch for ingestion for example.

---

This PR addresses the problems above by modifying the means by which we pull analytics records from Redis.

* new config option specifying the maximum number of records to pull from Redis at a time
* If PurgeChunk is 0 or unset, we will continue with current operation of pulling all analytics from Redis.
* If PurgeChunk is 100, this means we will pull 100 records at a time from Redis, meaning that we have more control over the chunk-size that we ship to upstream data-sources in one go.
* If PurgeChunk > 0, (enabled), then we check StorageExpirationTime. If not set, default to 60s. StorageExpirationTime will reset the TTL on the analytics record to StorageExpirationTime after a chunk has been purged. This means, that if the pump is not able to purge any records within 60s, then the analytics inside redis will expire. Effectively moving responsibility for resetting TTL on analytics record from the Gateway to the Pump - the entity which is purging the records.

## Considerations:

1. If data-sink is particularly slow, then there is a chance that the TTL inside Redis times out, and All analytics for that period will expire. This is expected functionality - because we don't want too many records building up in Redis - causing pressure on the DB.

2. Rather than TTL functionality inside Redis - maybe it would be more sensible / offer the option of controlling data-set size in Redis by using a kind of capped collection rather than a TTL?

3. Used in combination with PurgeDelay, We can now control & have visibility of how many RPS a single pump will purge.

```
PurgeDelay: 5
PurgeChunk: 10000
```

Means that the pump is configured to ingest & ship 2000 rps worth of analytics. And the data-sink(s) needs to be able to injest batches of 10k records. It becomes significantly easier to reason about & provides more guidance on configuration & sizing for pump deployments.

```
--- PASS: TestRedisClusterStorageManager_GetAndDeleteSet (0.08s)
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[]_0} (0.00s)
        redis_test.go:74: iterations 1 tt.in 0 tt.chunk 0
        redis_test.go:81: ---> 0: []
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one]_0} (0.00s)
        redis_test.go:74: iterations 1 tt.in 1 tt.chunk 0
        redis_test.go:81: ---> 0: [one]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two]_0} (0.00s)
        redis_test.go:74: iterations 1 tt.in 2 tt.chunk 0
        redis_test.go:81: ---> 0: [one two]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three]_0} (0.00s)
        redis_test.go:74: iterations 1 tt.in 3 tt.chunk 0
        redis_test.go:81: ---> 0: [one two three]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three_four]_0} (0.00s)
        redis_test.go:74: iterations 1 tt.in 4 tt.chunk 0
        redis_test.go:81: ---> 0: [one two three four]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three_four_five]_0} (0.00s)
        redis_test.go:74: iterations 1 tt.in 5 tt.chunk 0
        redis_test.go:81: ---> 0: [one two three four five]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[]_1} (0.00s)
        redis_test.go:74: iterations 0 tt.in 0 tt.chunk 1
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one]_1} (0.00s)
        redis_test.go:74: iterations 1 tt.in 1 tt.chunk 1
        redis_test.go:81: ---> 0: [one]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two]_1} (0.00s)
        redis_test.go:74: iterations 2 tt.in 2 tt.chunk 1
        redis_test.go:81: ---> 0: [one]
        redis_test.go:81: ---> 1: [two]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three]_1} (0.01s)
        redis_test.go:74: iterations 3 tt.in 3 tt.chunk 1
        redis_test.go:81: ---> 0: [one]
        redis_test.go:81: ---> 1: [two]
        redis_test.go:81: ---> 2: [three]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three_four]_1} (0.01s)
        redis_test.go:74: iterations 4 tt.in 4 tt.chunk 1
        redis_test.go:81: ---> 0: [one]
        redis_test.go:81: ---> 1: [two]
        redis_test.go:81: ---> 2: [three]
        redis_test.go:81: ---> 3: [four]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three_four_five]_1} (0.01s)
        redis_test.go:74: iterations 5 tt.in 5 tt.chunk 1
        redis_test.go:81: ---> 0: [one]
        redis_test.go:81: ---> 1: [two]
        redis_test.go:81: ---> 2: [three]
        redis_test.go:81: ---> 3: [four]
        redis_test.go:81: ---> 4: [five]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[]_2} (0.00s)
        redis_test.go:74: iterations 0 tt.in 0 tt.chunk 2
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one]_2} (0.00s)
        redis_test.go:74: iterations 1 tt.in 1 tt.chunk 2
        redis_test.go:81: ---> 0: [one]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two]_2} (0.00s)
        redis_test.go:74: iterations 1 tt.in 2 tt.chunk 2
        redis_test.go:81: ---> 0: [one two]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three]_2} (0.00s)
        redis_test.go:74: iterations 2 tt.in 3 tt.chunk 2
        redis_test.go:81: ---> 0: [one two]
        redis_test.go:81: ---> 1: [three]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three_four]_2} (0.00s)
        redis_test.go:74: iterations 2 tt.in 4 tt.chunk 2
        redis_test.go:81: ---> 0: [one two]
        redis_test.go:81: ---> 1: [three four]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three_four_five]_2} (0.01s)
        redis_test.go:74: iterations 3 tt.in 5 tt.chunk 2
        redis_test.go:81: ---> 0: [one two]
        redis_test.go:81: ---> 1: [three four]
        redis_test.go:81: ---> 2: [five]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[]_3} (0.00s)
        redis_test.go:74: iterations 0 tt.in 0 tt.chunk 3
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one]_3} (0.00s)
        redis_test.go:74: iterations 1 tt.in 1 tt.chunk 3
        redis_test.go:81: ---> 0: [one]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two]_3} (0.00s)
        redis_test.go:74: iterations 1 tt.in 2 tt.chunk 3
        redis_test.go:81: ---> 0: [one two]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three]_3} (0.00s)
        redis_test.go:74: iterations 1 tt.in 3 tt.chunk 3
        redis_test.go:81: ---> 0: [one two three]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three_four]_3} (0.00s)
        redis_test.go:74: iterations 2 tt.in 4 tt.chunk 3
        redis_test.go:81: ---> 0: [one two three]
        redis_test.go:81: ---> 1: [four]
    --- PASS: TestRedisClusterStorageManager_GetAndDeleteSet/in:_{[one_two_three_four_five]_3} (0.01s)
        redis_test.go:74: iterations 2 tt.in 5 tt.chunk 3
        redis_test.go:81: ---> 0: [one two three]
        redis_test.go:81: ---> 1: [four five]
```

### CODE REVIEW

File: storage/redis.go [hunk 1]
```diff
@@ -332,8 +332,8 @@ func (r *RedisClusterStorageManager) GetAndDeleteSet(keyName string) []interface
 
 	var lrange *redis.StringSliceCmd
 	_, err := r.db.TxPipelined(func(pipe redis.Pipeliner) error {
-		lrange = pipe.LRange(fixedKey, 0, -1)
-		pipe.Del(fixedKey)
+		lrange = pipe.LRange(fixedKey, 0, chunkSize-1)
+		pipe.LTrim(fixedKey, chunkSize, -1)
```
Comment (buger) [inline storage/redis.go]: Are you sure that in case, if the chunk is not set, `LTrim(fixedKey, -1, -1)` is equal to `Del` call?

Comment (asoorm) [reply storage/redis.go]: You are right - it misses the last one... will re-work it a bit.

Comment (asoorm) [reply storage/redis.go]: @buger PTAL

**Comment by buger** [comment]
Looks good, lets make it part of 2.9.4, e.g. part of the patch released after the next one.

File: storage/redis.go [hunk 2]
```diff
@@ -334,8 +334,13 @@ func (r *RedisClusterStorageManager) GetAndDeleteSet(keyName string) []interface
...
 	var lrange *redis.StringSliceCmd
 	_, err := r.db.TxPipelined(func(pipe redis.Pipeliner) error {
-		lrange = pipe.LRange(fixedKey, 0, -1)
-		pipe.Del(fixedKey)
+		lrange = pipe.LRange(fixedKey, 0, chunkSize-1)
+
+		if chunkSize == 0 {
+			pipe.Del(fixedKey)
+		} else {
+			pipe.LTrim(fixedKey, chunkSize, -1)
```
Comment (buger) [inline storage/redis.go]: @asoorm I just realized it won't work with gateway storage expiration mechanism. 
When we doing "Del" we remove the key, and its expiration. 
If we do just LTrim, there is a chance that analytics record will expire with active data.

Comment (asoorm) [reply storage/redis.go]: In which case, it should be the pump that is responsible for extending TTL to storage_expiration_time after a successful LTRIM and not the gateway.

**Comment by asoorm** [comment]
I've modified it a little - if the pump has PurgeChunk enabled, then it is now the pump's responsibility to extend the TTL after a successful purge.

TODO: Fix case insensitive import collision.

```
go install && tyk-pump                                                                  
../../../../pkg/mod/github.com/x-cray/logrus-prefixed-formatter@v0.5.2/formatter.go:14:2: case-insensitive import collision: "github.com/sirupsen/logrus" and "github.com/Sirupsen/logrus"
```

**Comment by letzya** [comment]
so now it's ready for review?

**Comment by asoorm** [comment]
@buger bump

File: storage/redis.go (hunk 2 above)
Comment (tbuchaillot) [reply storage/redis.go]: @asoorm @buger I've been playing with this and it seems that pump expiration > gateway expiration. 
I'm running this with a low `purge_chunk`, default `storage_expiration_time` (60s), and constants requests against an API and it seems that the analytics key in the Redis keeps updating the TTL to the `storage_expiration_time` so I don't think we have a problem with this.

**Review by tbuchaillot — APPROVED** [review]
LGTM

**Review by matiasinsaurralde — APPROVED** [review]
LGTM

**Pull request merged**